includes/class-wc-payments-redirect-service.php

```diff
@@ -135,11 +136,16 @@ public function redirect_to_connect_page( ?string $error_message = null, string
 		];
 
 		if ( count( $params ) === count( array_intersect_assoc( $_GET, $params ) ) ) { // phpcs:disable WordPress.Security.NonceVerification.Recommended
-			// We are already in the onboarding page, do nothing.
+			// We are already on the Connect page. Do nothing.
 			return;
 		}
 
-		if ( '' !== $from ) {
+		// If we were not given a source, try to get it from the request URL.
+		if ( ! isset( $from ) && isset( $_GET['from'] ) ) {
```

Comment:
could you please add unit test for this?

Reply:
Done in https://github.com/Automattic/woocommerce-payments/pull/9036/commits/298523d5466edd87e74bb75530dca85bbacad25d